### PR TITLE
Keys Backup can be setup twice #9510

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #9510

Check for existing backup in the last step of the setup.
Check remote state when VectorHomeActivity is resumed to update banner.

image

### CODE REVIEW

File: vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt [hunk 1]
```diff
@@ -189,7 +184,45 @@ class KeysBackupSetupSharedViewModel : ViewModel() {
...
+
+        keysBackup.getCurrentVersion(object : SimpleApiCallback<KeysVersionResult?>(failureCallBack) {
+            override fun onSuccess(info: KeysVersionResult?) {
+                loadingStatus.value = null
+                if (info?.version.isNullOrBlank()) {
+                    //should not happen
+                    processOnCreate()
+                } else {
+                    //we should prompt
+                    AlertDialog.Builder(context)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt]: The dialog should not be in the model

File: vector/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -1294,6 +1294,11 @@ Why choose Riot.im?
     <string name="keys_backup_setup_step3_save_button_title">Save as File</string>
     <string name="recovery_key_export_saved_as_warning">The recovery key has been saved to \'%s\'.\n\nWarning: this file may be deleted if the application is uninstalled.</string>
 
+    <string name="keys_backup_setup_override_backup_prompt_tile">A backup already exist in this location</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml]: A backup already exist**s** ~~in this location~~ **on your homeserver**

File: vector/src/main/java/im/vector/activity/VectorHomeActivity.java [hunk 3]
```diff
@@ -666,6 +666,11 @@ public void onClick(DialogInterface dialog, int which) {
         addBadgeEventsListener();
 
         checkNotificationPrivacySetting();
+
+        //Force remote backup state update to update the banner if needed
+        if (mSession.getCrypto() != null) {
+            mSession.getCrypto().getKeysBackup().checkAndStartKeysBackup();
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/VectorHomeActivity.java]: Is it done elsewhere?

File: vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt]: done

File: vector/src/main/res/values/strings.xml (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/res/values/strings.xml]: done

File: vector/src/main/java/im/vector/activity/VectorHomeActivity.java (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/VectorHomeActivity.java]: as discussed

File: vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt [hunk 4]
```diff
@@ -190,30 +203,22 @@ class KeysBackupSetupSharedViewModel : ViewModel() {
         keysBackup.getCurrentVersion(object : SimpleApiCallback<KeysVersionResult?>(failureCallBack) {
             override fun onSuccess(info: KeysVersionResult?) {
                 loadingStatus.value = null
-                if (info?.version.isNullOrBlank()) {
+                if (info?.version.isNullOrBlank() || forceOverride) {
                     //should not happen
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt]: Maybe remove this comment

File: vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt [hunk 5]
```diff
@@ -189,7 +197,37 @@ class KeysBackupSetupSharedViewModel : ViewModel() {
...
+
+        keysBackup.getCurrentVersion(object : SimpleApiCallback<KeysVersionResult?>(failureCallBack) {
+            override fun onSuccess(info: KeysVersionResult?) {
+                loadingStatus.value = null
+                if (info?.version.isNullOrBlank() || forceOverride) {
+                    //should not happen
+                    processOnCreate()
+                } else {
+                    //we should prompt
+                    loadingStatus.value = null
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt]: This line is not necessary (it's done line 205)

File: vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt [hunk 6]
```diff
@@ -189,7 +197,37 @@ class KeysBackupSetupSharedViewModel : ViewModel() {
...
+                    isCreatingBackupVersion.value = false
+                    navigateEvent.value = LiveEvent(NAVIGATE_PROMPT_REPLACE)
+
+                }
+            }
+
+            fun processOnCreate() {
+                keysBackup.createKeysBackupVersion(megolmBackupCreationInfo!!, object : SimpleApiCallback<KeysVersion>(failureCallBack) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keysbackup/setup/KeysBackupSetupSharedViewModel.kt]: The loading status is not set to true before this call?

File: vector/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -1294,6 +1294,11 @@ Why choose Riot.im?
     <string name="keys_backup_setup_step3_save_button_title">Save as File</string>
     <string name="recovery_key_export_saved_as_warning">The recovery key has been saved to \'%s\'.\n\nWarning: this file may be deleted if the application is uninstalled.</string>
 
+    <string name="keys_backup_setup_override_backup_prompt_tile">A backup already exist on your HomeServer</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml]:
```suggestion
    <string name="keys_backup_setup_override_backup_prompt_tile">A backup already exists on your homeserver</string>
```

File: vector/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -1294,6 +1294,11 @@ Why choose Riot.im?
...
+    <string name="keys_backup_setup_override_backup_prompt_description">It looks like you already have setup key backup from another device. Do you want to replace it with the one you’re creating?</string>
+    <string name="keys_backup_setup_override_replace">Replace</string>
+    <string name="keys_backup_setup_override_stop">Stop</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml]:
```suggestion
    <string name="keys_backup_setup_override_stop">Cancel</string>
```

**Pull request merged**